### PR TITLE
Add gRPC connection monitor and fix related bug

### PR DESCRIPTION
**Pull request by Smityz**
Signed-off-by: Smityz [smityz@qq.com](mailto:smityz@qq.com)

fix: https://github.com/tikv/client-go/issues/522 https://github.com/tikv/client-go/issues/653

After a tikv server restart, the grpc connection state on the client becomes IDLE and needs a request to be sent through this connection or a manual connect to change it to the READY state. In the current implementation, I believe trying to`Connect()` when the state is IDLE is a relatively small modification.

And add a monitor of grpc connection state can help us to find related bugs

Screenshot 2023-01-11 at 18 11 38

### CODE REVIEW

**Comment by disksing** [comment]
@hicqu @MyonKeminta Please take a look.

File: internal/client/client.go [hunk 1]
```diff
@@ -114,30 +114,119 @@ type connArray struct {
...
+			go c.start()
+		},
+	)
+}
+
+func (c *connMonitor) Stop() {
+	c.stopOnce.Do(
+		func() {
+			if c.stop != nil {
+				close(c.stop)
```
Comment (hicqu) [inline internal/client/client.go:175]: Should set stop to nil here?

Comment (Smityz) [reply internal/client/client.go:175]: `stop` is a chan here, why should set it to nil?

Comment (Smityz) [reply internal/client/client.go:175]: `stopOnce` can  promise

**Comment by disksing** [comment]
@hicqu any more comments?

**Review by hicqu — APPROVED** [review]

**Pull request merged**